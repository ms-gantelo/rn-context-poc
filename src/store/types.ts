import { IArtistState } from './artist/state';
import { IAlbumState } from './album/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IGlobalState = {
  artist: IArtistState;
  album: IAlbumState;
};

export type IReducer<S = any, A = any> = (state: S, action: IAction<A>) => S;

export interface Deps {
  apiService: { request: Function };
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
}
