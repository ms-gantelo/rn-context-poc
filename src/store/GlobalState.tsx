import React, { createContext, memo, PropsWithChildren, useReducer, Dispatch, useEffect } from 'react';

import { initialState as artistInitialState } from './artist/state';
import { reducer as artistReducer } from './artist/reducer';
import { initialState as albumInitialState } from './album/state';
import { reducer as albumReducer } from './album/reducer';
import { IAction, IGlobalState, Deps, IReducer } from './types';

export const initialState: IGlobalState = {
  artist: artistInitialState,
  album: albumInitialState
};

export function combineReducers<GS>(reducerMap: Record<string, (state: any, action: IAction) => any>): IReducer<GS> {
  return (gState: GS, gAction: IAction) => {
    const nextState = Object.entries(reducerMap).reduce(
      (total, [stateKey, reducer]) => ({ ...total, [stateKey]: reducer(gState[stateKey], gAction) }),
      {} as GS
    );
    const hasChanged = !!Array.from(new Set([...Object.keys(gState), ...Object.keys(nextState)])).find(key => gState[key] !== nextState[key]);
    return hasChanged ? nextState : gState;
  };
}

export const GlobalContext = createContext<{ state: IGlobalState; dispatch: Dispatch<any>; deps: Deps }>({
  state: initialState,
  dispatch: () => null,
  deps: null
});

export const GlobalProvider = memo(
  ({
    children,
    deps,
    initState = initialState,
    combinedReducer = combineReducers<IGlobalState>({ artist: artistReducer, album: albumReducer })
  }: PropsWithChildren<{ deps: Deps; initState?: IGlobalState; combinedReducer?: IReducer<IGlobalState> }>) => {
    const [state, dispatch] = useReducer(combinedReducer, initState);
    useEffect(() => deps.stateSnapshot.set(state), [state]);
    return <GlobalContext.Provider value={{ state, dispatch, deps }}>{children}</GlobalContext.Provider>;
  }
);
