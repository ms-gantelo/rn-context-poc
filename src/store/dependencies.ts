import { Deps, IGlobalState } from './types';

export const getDeps = (): Deps => ({
  apiService: { request: fetch },
  stateSnapshot: (() => {
    let state: IGlobalState;
    return {
      set: (s: IGlobalState) => {
        state = s;
      },
      get: () => state
    };
  })()
});
