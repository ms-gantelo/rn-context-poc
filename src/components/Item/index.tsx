import React, { memo, useContext } from 'react';
import { ScrollView, Text, Image, View } from 'react-native';

import { GlobalContext } from '../../store/GlobalState';

export interface Props {
  route: any; // RouteProp<{ id: string; }, any>; /** @todo find out how to use this type */
}

const Item = ({ route }: Props) => {
  const { state } = useContext(GlobalContext);
  const artist = state.artist.artistMap[route?.params?.id];
  return (
    <ScrollView contentInsetAdjustmentBehavior="automatic">
      <Image style={{ width: '100%', height: 100 }} source={{ uri: artist.strArtistBanner }} />
      <View style={{ padding: 15 }}>
        <Text>{artist.strArtist}</Text>
        <Text ellipsizeMode="head" numberOfLines={30}>
          {artist.strBiographyEN}
        </Text>
      </View>
    </ScrollView>
  );
};

export default memo(Item);
