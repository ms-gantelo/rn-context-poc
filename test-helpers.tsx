import { IArtist } from './src/store/artist/state';
import { Deps } from './src/store/types';

export const getMockDeps = (): Deps =>
  ({
    apiService: { request: jest.fn().mockResolvedValue({ json: () => ({}) }) },
    stateSnapshot: { set: jest.fn(), get: jest.fn().mockReturnValue({}) }
  } as any);

export const getArtist_1 = (): IArtist => ({
  idArtist: '1',
  intBornYear: 1970,
  strArtist: 'Freddie Mercuty',
  strArtistThumb: 'thumb.png',
  strArtistBanner: 'banner.png',
  strArtistFanart: 'fanart.png',
  strBiographyEN: 'this is a bio',
  strCountry: 'USA',
  strStyle: 'rock'
});
