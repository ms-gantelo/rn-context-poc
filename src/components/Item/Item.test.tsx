import React from 'react';
import { render } from 'react-native-testing-library';

import Item from '.';
import { getArtist_1, getMockDeps } from '../../../test-helpers';
import { GlobalProvider, initialState } from '../../store/GlobalState';

describe('Item', () => {
  it('should render', () => {
    const { toJSON } = render(
      <GlobalProvider
        deps={getMockDeps()}
        initState={{ ...initialState, artist: { ...initialState.artist, artistMap: { [getArtist_1().idArtist]: getArtist_1() } } }}
      >
        <Item route={{ params: { id: getArtist_1().idArtist } }} />
      </GlobalProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
