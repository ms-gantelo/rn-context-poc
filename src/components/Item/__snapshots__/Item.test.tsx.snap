// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render 1`] = `
<RCTScrollView
  contentInsetAdjustmentBehavior="automatic"
>
  <View>
    <Image
      source={
        Object {
          "uri": "banner.png",
        }
      }
      style={
        Object {
          "height": 100,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "padding": 15,
        }
      }
    >
      <Text>
        Freddie Mercuty
      </Text>
      <Text
        ellipsizeMode="head"
        numberOfLines={30}
      >
        this is a bio
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
